backend/routes/auth.js

```diff
@@ -52,5 +54,45 @@ router.post("/signup", async (req, res) => {
   res.status(201).json(newUser);
 });
 
+// login routes!
+
+const rateLimit = require("express-rate-limit");
+
+const loginLimiter = rateLimit({
+  windowMs: 15 * 60 * 1000, // 15 minutes
+  max: 100, // Limit each IP to 10 login attempts per windowMs
+  message: { error: "Too many failed login attempts. Try again in 10 minutes" },
+});
+
+router.post("/login", loginLimiter, async (req, res) => {
+  const { username, password } = req.body;
+  const user = await prisma.user.findUnique({
+    where: { username },
+  });
+  if (!user) {
+    return res.status(400).json({ error: "Invalid username or password." });
+  }
+
+  const isValidPassword = await argon2.verify(user.password, password);
+  if (!isValidPassword) {
+    return res.status(400).json({ error: "Invalid username or password." });
+  }
+  req.session.userId = user.id;
+  res.status(200).json({ message: "Login successful!" });
+});
+
+// check if valid session exists!
+router.get("/me", async (req, res) => {
+  const userId = req.session.userId;
+  if (userId == null) {
+    return res.status(401).json({ message: "Not logged in" });
+  }
+  const user = await prisma.user.findUnique({
+    where: { id: userId },
+    select: { username: true }, // Only return necessary data
+  });
+
+  res.json({ id: req.session.userId, username: user.username });
```

Comment:
If this is just existence check, why do you need to return `username`? Is it used on the client side?